File: php/src/Advertisement/Domain/Errors/InvalidEmailFormatError.php
<?php
declare(strict_types=1);

namespace Demo\App\Advertisement\Domain\Errors;

use Demo\App\Common\Result;

final readonly class InvalidEmailFormatError
{
    public static function build(string $email): Result
    {
        return Result::failure('Invalid email format ' . $email);
    }
}

File: kotlin/src/main/kotlin/advertisement/domain/errors/InvalidEmailFormatError.kt
package advertisement.domain.errors

import common.domain.DomainError

open class InvalidEmailFormatError private constructor(errorMessage: String): DomainError(errorMessage) {
    companion object {
        fun withEmail(value: String): InvalidEmailFormatError {
            val errorMessage = "Invalid email format $value"

            return InvalidEmailFormatError(errorMessage)
        }
    }
}
